File: src/main/java/pcb/uwu/amounts/fundamental/Temperature.java
package pcb.uwu.amounts.fundamental;

import pcb.uwu.core.AbstractUnitAmount;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.fundamental.TemperatureUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Temperature extends AbstractUnitAmount<TemperatureUnit> {

	// region constructors

	public Temperature(Number value, TemperatureUnit unit) {
		super(value, unit);
	}

	public Temperature(Number value, Magnitude magnitude, TemperatureUnit unit) {
		super(value, magnitude, unit);
	}

	public Temperature(String value, TemperatureUnit unit) {
		super(value, unit);
	}

	public Temperature(String value, Magnitude magnitude, TemperatureUnit unit) {
		super(value, magnitude, unit);
	}

	public Temperature(BigDecimal value, TemperatureUnit unit) {
		super(value, unit);
	}

	public Temperature(BigDecimal value, Magnitude magnitude, TemperatureUnit unit) {
		super(value, magnitude, unit);
	}

	public Temperature(BigDecimalAmount amount, TemperatureUnit unit) {
		super(amount, unit);
	}

	public Temperature(BigDecimalAmount amount, Magnitude magnitude, TemperatureUnit unit) {
		super(amount, magnitude, unit);
	}


	// endregion

	// region implement UnitAmount

	@Override
	public Temperature plus(UnitAmount<TemperatureUnit> other, MathContext mathContext) {
		return new Temperature(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Temperature minus(UnitAmount<TemperatureUnit> other, MathContext mathContext) {
		return new Temperature(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Temperature multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Temperature(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Temperature dividedBy(BigDecimal other, MathContext mathContext) {
		return new Temperature(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Temperature convertTo(TemperatureUnit unit) {
		return new Temperature(getAmountIn(unit), unit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/base/Temperature.kt
package pcb.uwu.amount.base

import pcb.uwu.core.Amount
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.TemperatureUnit
import pcb.uwu.util.UnitAmountUtils
import java.math.BigDecimal

open class Temperature : CompositeUnitAmount<TemperatureUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: TemperatureUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: TemperatureUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(value: BigDecimal,
                magnitude: Magnitude = NATURAL,
                unit: TemperatureUnit)
            : super(value, magnitude, unit)

    @JvmOverloads
    constructor(amount: Amount,
                magnitude: Magnitude = NATURAL,
                unit: TemperatureUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(temperature: UnitAmount<TemperatureUnit>) =
        Temperature(amount = this.amount + (temperature to this.unit).amount,
                    unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(temperature: UnitAmount<TemperatureUnit>) =
        Temperature(amount = this.amount - (temperature to this.unit).amount,
                    unit = this.unit)

    override fun times(scalar: Number) =
        Temperature(amount = this.amount * scalar,
                    unit = this.unit)

    override fun div(scalar: Number) =
        Temperature(amount = this.amount / scalar,
                    unit = this.unit)

    override fun to(unit: TemperatureUnit) =
        Temperature(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                    unit = unit)

    // endregion
}